### DISCUSSION TITLE
How to load dynamic libraries?

### DISCUSSION THREAD

**Discussion by 1oglop1** [cargo-lambda/cargo-lambda#504] [Q&A]
Hello @calavera, sorry for tagging you directly, but this is probably a low-level issue requiring specific knowledge.
I have lambda functions in multiple languages and I found this extension https://github.com/rails-lambda/crypteia

The extension `crypteia` can override the environment variables with SSM parameters, similarly to how it works in ECS. 

Eg. `MYVAR=x-crypteia-ssm:/my_ssm_param` is changed via runtime to `MYVAR=value`

The extension works for several languages, Node, Python, ... 
It contains 2 parts `crypteia` binary and the library `libcrypteia.so`. 

Then for running Node.js lamdba all I needed was to set `LD_PRELOAD=path/to/libcrypteia.so` however this had no effect on the lambda I built with Rust.

Do you have any idea what am I missing?

**Comment by calavera** [comment]
I have no idea how that library works. It's probably better if you discuss it with the developers that built it.

**Reply by 1oglop1** [reply]
The problem is not the library, it does what it should do. 
The only time it does not do what it is supposed to do is with the Rust binary. 
Originally I thought it might be something with the cargo-lambda zig compiler, but then I compiled the lambda with a cargo compiler and got the same result. 

The only thing I haven't tried yet was running the lambda runtime binary inside the local container and adding the extension layer. 

I was hoping you might have some insights or tips on how to debug this because I am not so familiar with these low-level things yet.

The extension overrides C function `getenv` (I assume from glibC) 

However, there seem to be 2 different libraries providing access to Env vars.

one of them is GLIB the other one LibC - I do not understand differences yet. 

but here are their rust versions
https://docs.rs/libc/latest/libc/index.html
https://docs.rs/glib/latest/glib/index.html

**Reply by calavera** [reply]
> The extension overrides C function getenv

This is exactly why you should ask the people that built the extension. I don't know why or how they do that. They should probably give you an example for how to use their extension with Rust projects.
